### DISCUSSION TITLE
VerifyTestVectors in CKKS

### DISCUSSION THREAD

**Discussion by Maokami** [tuneinsight/lattigo#424] [Q&A] [answered]
Hello, Lattigo Team!

May I ask what is the purpose of the VerifyTestVectors in CKKS test?
https://github.com/tuneinsight/lattigo/blob/1f00f3ca7147a13f1313c64610385a904f84db7e/schemes/ckks/precision.go#L70-L93

Specifically, from the following codes, it seems that the CKKS test codes want to verify whether `mean(|(m1+m2) - Dec(Add(Enc(m1+m2)))|) > LogDefaultScale - LogN - 2`.
https://github.com/tuneinsight/lattigo/blob/1f00f3ca7147a13f1313c64610385a904f84db7e/schemes/ckks/precision.go#L83
https://github.com/tuneinsight/lattigo/blob/1f00f3ca7147a13f1313c64610385a904f84db7e/schemes/ckks/ckks_test.go#L426-L441

I wonder why this property must be checked, and is there any theoretical explanation of this?
Thanks!

**Comment by Pro7ech** [comment, marked as answer]
The CKKS scheme is approximate, we must therefore define some lower bound on the accepted precision when checking the correctness of the operations. This defines heuristic lower bound that is a function of the ring degree and scaling factor and is achieved with overwhelming probability.

**Reply by Maokami** [reply]
Thank you, @Pro7ech!

Is the _heuristic lower bound_ calculated based on CKKS papers, such as the original CKKS paper and its [reduced error version](https://eprint.iacr.org/2020/1118.pdf)?
